### PR TITLE
[quant][pt2e] Refactor insert observer to do sharing checking in the same place

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113500
* __->__ #113458

Summary:
Previously it is scatter in two different places: before inserting observer and during observer,
this PR moved everything before we insert observer

* Next: refactor QuantizationSpec and check more fields for sharing

Test Plan:
CI (regression tests)

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/pt2e/prepare.py [hunk 1]
```diff
@@ -179,21 +198,23 @@ def _get_edge_or_node_to_group_id(edge_or_node_to_qspec: Dict[EdgeOrNode, Quanti
...
-                    # since dtype is the same (we may want to extend this to be a more strict check
-                    # in the future)
-                    # so we point from `input_edge` to `arg` (output of the argument)
-                    _union(arg, input_edge, shared_with_map)
+                # sharing with previous output
+                _union_input_edge_with(input_edge, input_edge_root_qspec, arg, edge_or_node_to_qspec, shared_with_map)
+
+                # sharing with other users of the previous output
+                # (arg, user)
+                for user in arg.users:
```
Comment (kimishpatel) [inline torch/ao/quantization/pt2e/prepare.py:206]: This is not a refactor change, right?

**Review by kimishpatel — APPROVED** [review]
left a comment

File: torch/ao/quantization/pt2e/prepare.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/prepare.py:206]: this is a refactor, we remove this logic from insert observers and added it here

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but it is failing executorch export test for llama2' -c ghfirst

The error is:

```
======================================================================
ERROR: test_llama_qnn (executorch.examples.models.llama2.fb.test_llama_qnn.LlamaLiteModelExportTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/executorch/examples/models/llama2/fb/test_llama_qnn.py", line 70, in test_llama_qnn
    prepared_gm = prepare_pt2e(captured_gm, QnnHtpQuantizer())
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/quantize_pt2e.py", line 109, in prepare_pt2e
    model = prepare(model, node_name_to_scope, is_qat=False)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/pt2e/prepare.py", line 442, in prepare
    edge_or_node_to_group_id = _get_edge_or_node_to_group_id(edge_or_node_to_qspec)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/pt2e/prepare.py", line 206, in _get_edge_or_node_to_group_id
    for user in arg.users:
AttributeError: 'int' object has no attribute 'users'

----------------------------------------------------------------------
Ran 1 test in 6.339s
```

Just FYI, I'm working on bring this test to OSS https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L1014-L1016

**Comment by jerryzh168** [comment]
> @pytorchbot revert -m 'Sorry for reverting your change but it is failing executorch export test for llama2' -c ghfirst
> 
> The error is:
> 
> ```
> ======================================================================
> ERROR: test_llama_qnn (executorch.examples.models.llama2.fb.test_llama_qnn.LlamaLiteModelExportTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/executorch/examples/models/llama2/fb/test_llama_qnn.py", line 70, in test_llama_qnn
>     prepared_gm = prepare_pt2e(captured_gm, QnnHtpQuantizer())
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/quantize_pt2e.py", line 109, in prepare_pt2e
>     model = prepare(model, node_name_to_scope, is_qat=False)
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/pt2e/prepare.py", line 442, in prepare
>     edge_or_node_to_group_id = _get_edge_or_node_to_group_id(edge_or_node_to_qspec)
>   File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/24c04fc3eb9063ab/executorch/examples/models/llama2/fb/__test_llama_qnn__/test_llama_qnn#link-tree/torch/ao/quantization/pt2e/prepare.py", line 206, in _get_edge_or_node_to_group_id
>     for user in arg.users:
> AttributeError: 'int' object has no attribute 'users'
> 
> ----------------------------------------------------------------------
> Ran 1 test in 6.339s
> ```
> 
> Just FYI, I'm working on bring this test to OSS https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L1014-L1016

this feels like an error on HtpQnnQuantizer side actually, I'll check with Shen

**Comment by jerryzh168** [comment]
reland PR: https://github.com/pytorch/pytorch/pull/113920

**Pull request closed without merging**